### PR TITLE
Create Force Unlock Messages

### PR DESCRIPTION
**Pull request by mattverse**
Closes: #2727 

## What is the purpose of the change

This PR creates a msg for force unlocking for the addresses that has been passed via governance and registered as params in the lockup keeper.

This PR also includes adding additional logic to the lockup keeper since we did not support partial unlockings for locks that are already  unlocking. ForceUnlock allows force unlocking of locks that are currently being unlocked.

## Brief Changelog

- Create ForceUnlock message for lockup keeper

## Testing and Verifying

Testing WIP: Planning to do PR it separately as test cases are getting big, would prefer separate review for the test cases / edge cases dealt within the test cases

Some of the tests being worked on:
- Test that is assets being superfluid delegated, undelegated
- Test partial force unlocking

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: proto/osmosis/lockup/tx.proto [hunk 1]
```diff
@@ -71,3 +72,15 @@ message MsgExtendLockup {
 }
 
 message MsgExtendLockupResponse { bool success = 1; }
+
+message MsgForceUnlock {
```
Comment (alexanderbez) [inline proto/osmosis/lockup/tx.proto:76]: nit: add docs for request and response types :)

File: proto/osmosis/lockup/tx.proto [hunk 2]
```diff
@@ -71,3 +72,15 @@ message MsgExtendLockup {
...
+  string owner = 1 [ (gogoproto.moretags) = "yaml:\"owner\"" ];
```
Comment (alexanderbez) [inline proto/osmosis/lockup/tx.proto:77]: Is this an SDK address? If so, shouldn't we add `(cosmos_proto.scalar) = "cosmos.AddressString"`? Maybe our proto tooling doesn't support this yet?

Comment (alexanderbez) [reply proto/osmosis/lockup/tx.proto:77]: Also, `  option (cosmos.msg.v1.signer) = "from_address";`

File: x/lockup/keeper/msg_server.go [hunk 3]
```diff
@@ -182,3 +182,43 @@ func (server msgServer) ExtendLockup(goCtx context.Context, msg *types.MsgExtend
...
+
+	// force unlock given lock
+	// This also supports the case of force unlocking lock as a whole when msg.Coins
+	// provided is empty.
+	err = server.keeper.PartialForceUnlock(ctx, *lock, msg.Coins)
+	if err != nil {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, err.Error())
+	}
+
+	return &types.MsgForceUnlockResponse{Success: true}, nil
```
Comment (alexanderbez) [inline x/lockup/keeper/msg_server.go:223]: When would we ever return false? Is it worth having the `Success` field? Are there other fields we might want?

**Review by alexanderbez — APPROVED** [review]
LGTM! I can't speak much to the `splitLock` implementation, but the utilization looks sound.

File: app/upgrades/v12/upgrades.go [hunk 4]
```diff
@@ -83,6 +84,7 @@ func CreateUpgradeHandler(
 
 		// Set TWAP parameters to default values.
 		keepers.TwapKeeper.SetParams(ctx, twaptypes.DefaultParams())
+		keepers.LockupKeeper.SetParams(ctx, lockuptypes.DefaultParams())
```
Comment (p0mvn) [inline app/upgrades/v12/upgrades.go:87]: We should probably make a v13 upgrade handler and add it there

File: x/lockup/types/params.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	return paramtypes.NewKeyTable().RegisterParamSet(&Params{})
+}
+
+func NewParams(forceUnlockAllowedAddresses []string) Params {
+	return Params{
+		ForceUnlockAllowedAddresses: forceUnlockAllowedAddresses,
+	}
+}
+
+// default lockup module parameters.
```
Comment (p0mvn) [inline x/lockup/types/params.go:28]:
```suggestion
// DefaultParams returns default lockup module parameters.
```

File: x/lockup/types/msgs.go [hunk 6]
```diff
@@ -165,3 +166,33 @@ func (m MsgExtendLockup) GetSigners() []sdk.AccAddress {
...
+		Owner: owner.String(),
+		ID:    id,
+		Coins: coins,
+	}
+}
+
+func (m MsgForceUnlock) Route() string { return RouterKey }
+func (m MsgForceUnlock) Type() string  { return TypeMsgBeginUnlockingAll }
+func (m MsgForceUnlock) ValidateBasic() error {
+	_, err := sdk.AccAddressFromBech32(m.Owner)
```
Comment (p0mvn) [inline x/lockup/types/msgs.go:184]: Should we also validate id and coins here?

File: x/lockup/keeper/keeper.go [hunk 7]
```diff
@@ -17,21 +18,44 @@ type Keeper struct {
...
+
+// SetParams sets the total set of lockup parameters.
+func (k Keeper) SetParams(ctx sdk.Context, params types.Params) {
+	k.paramSpace.SetParamSet(ctx, &params)
+}
+
+func (k Keeper) GetForceUnlockAllowedAddresses(ctx sdk.Context) (forceUnlockAllowedAddresses []string) {
+	return k.GetParams(ctx).ForceUnlockAllowedAddresses
+}
+
```
Comment (p0mvn) [inline x/lockup/keeper/keeper.go:58]: non-blocking nit: I'm generally unsure if this method is needed if we can get the addresses from params

File: x/lockup/keeper/msg_server.go [hunk 8]
```diff
@@ -182,3 +182,43 @@ func (server msgServer) ExtendLockup(goCtx context.Context, msg *types.MsgExtend
 
 	return &types.MsgExtendLockupResponse{}, nil
 }
+
+// ForceUnlock ignores unlock duration and immediately unlocks the lock.
+// This message is only allowed for governance passed accounts that is kept as parameter in the lockup module.
```
Comment (p0mvn) [inline x/lockup/keeper/msg_server.go:187]:
```suggestion
// This message is only allowed for governance-passed accounts that are kept as parameter in the lockup module.
```

File: x/lockup/keeper/msg_server.go [hunk 9]
```diff
@@ -182,3 +182,43 @@ func (server msgServer) ExtendLockup(goCtx context.Context, msg *types.MsgExtend
...
+	// check for chain parameter that the address is allowed to force unlock
+	forceUnlockAllowedAddresses := server.keeper.GetForceUnlockAllowedAddresses(ctx)
+	found := false
+	for _, address := range forceUnlockAllowedAddresses {
+		if address == msg.Owner {
+			found = true
+		}
+	}
+	if !found {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrUnauthorized, "Sender not allowed to force unlock")
```
Comment (p0mvn) [inline x/lockup/keeper/msg_server.go:201]: consider formatting the sender address for ease of debugging in the future

**Review by p0mvn — APPROVED** [review]

File: proto/osmosis/lockup/tx.proto (hunk 2 above)
Comment (mattverse) [reply proto/osmosis/lockup/tx.proto:77]: what does `(cosmos_proto.scalar) = "cosmos.AddressString"` and `option (cosmos.msg.v1.signer) = "from_address";` do?

Orignial intention was to recieve it as string and then convert it to address

File: x/lockup/keeper/msg_server.go (hunk 3 above)
Comment (mattverse) [reply x/lockup/keeper/msg_server.go:223]: would return false on the response whenever the transaction fails due to error in msg server, I think this one field would suffice for the field of this response!

File: x/lockup/types/msgs.go (hunk 6 above)
Comment (mattverse) [reply x/lockup/types/msgs.go:184]: good catch! added

File: app/upgrades/v12/upgrades.go (hunk 4 above)
Comment (mattverse) [reply app/upgrades/v12/upgrades.go:87]: good catch! just created a v13 upgrade handler

File: x/lockup/keeper/keeper.go (hunk 7 above)
Comment (mattverse) [reply x/lockup/keeper/keeper.go:58]: I originally thought this looked messy(having to call params to call get the addresses), but on the second thought I tink that way makes everything more intuitive, reverting this part so that we call from params directly

File: x/lockup/keeper/msg_server.go (hunk 9 above)
Comment (mattverse) [reply x/lockup/keeper/msg_server.go:201]: added!

File: proto/osmosis/lockup/params.proto [hunk 10]
```diff
@@ -0,0 +1,12 @@
+syntax = "proto3";
+package osmosis.lockup;
+
+import "gogoproto/gogo.proto";
+
+option go_package = "github.com/osmosis-labs/osmosis/v12/x/lockup/types";
+
+// Params defines the parameters for the tokenfactory module.
```
Comment (ValarDragon) [inline proto/osmosis/lockup/params.proto:8]:
```suggestion
```

**Comment by apollo-sturdy** [comment]
Hi. Wondering if there are any updates to this issue? Mainly curious if the API is finalized so that we (Apollo DAO) can implement this in our vault contract.
